src/main/jni/JsonCollator/com_couchbase_lite_android_SQLiteJsonCollator.cpp

```diff
@@ -303,60 +283,22 @@ static int compareStringsUnicodeFast(const char** in1, const char** in2) {
     return 0;
 }
 
-static jstring createJavaStringFromJSON(const char** in) {
-    // Scan the JSON string to find its end and whether it contains escapes:
-    const char* start = ++*in;
-    unsigned escapes = 0;
-    const char* str;
-    for (str = start; *str != '"'; ++str) {
-        if (*str == '\\') {
-            ++str;
-            if (*str == 'u') {
-                escapes += 5;  // \uxxxx adds 5 bytes
-                str += 4;
-            } else
-                escapes += 1;
-        }
-    }
-    *in = str + 1;
-    size_t length = str - start;
-    
-    char* buf = NULL;
-    length -= escapes;
-    buf = (char*) malloc(length + 1);
-    char* dst = buf;
-    char c;
-    for (str = start; (c = *str) != '"'; ++str) {
-        if (c == '\\')
-            c = convertEscape(&str);
-        *dst++ = c;
-    }
-    *dst++ = 0; //null terminate
-    start = buf;
-    //LOGV("After stripping escapes string is: %s", start);
-    
-    JNIEnv *env = getEnv();
-    jstring result = env->NewStringUTF(start);
-    if (buf != NULL) {
-        free(buf);
-    }
-    if (result == NULL) {
-        LOGE("Failed to convert to string: start=%p, length=%u", start, length);
-    }
-    return result;
-}
-
 static int compareStringsUnicode(const char** in1, const char** in2) {
     int result = compareStringsUnicodeFast(in1, in2);
     if (result > -2)
         return result;
-    // Fast compare failed, so resort to using NSString:
-    jstring str1 = createJavaStringFromJSON(in1);
-    jstring str2 = createJavaStringFromJSON(in2);
-    JNIEnv *env = getEnv();
-    result = env->CallStaticIntMethod(collatorClass, compareMethod, str1, str2);
-    env->DeleteLocalRef(str1);
-    env->DeleteLocalRef(str2);
+
+    // ICU4C - Collator
+    if(coll == NULL){
+        UErrorCode status = U_ZERO_ERROR; 
+        coll = Collator::createInstance(locale, status);
+        if(!U_SUCCESS(status)) {
+            LOGE("Failed to create Collator instance: status=%d", status);
+            return -3;  
```

Comment:
What will happen in in CollateJSON() when -3 is returned? It's really not clear from reading CollateJSON() code about what will happen.